src/StackExchange.Redis/ChannelMessageQueue.cs

```diff
@@ -193,68 +218,63 @@ private async Task OnMessageAsyncImpl()
             while (!Completion.IsCompleted)
             {
                 ChannelMessage next;
-                try { if (!TryRead(out next)) next = await ReadAsync().ConfigureAwait(false); }
-                catch (ChannelClosedException) { break; } // expected
-                catch (Exception ex)
+
+                // Keep trying to read values
+                while (!TryRead(out next))
                 {
-                    _parent.multiplexer?.OnInternalError(ex);
-                    break;
+                    // If we fail, wait for an item to appear
+                    if (!await _queue.WaitToReadAsync())
+                    {
+                        // Channel is closed
+                        break;
+                    }
+
+                    // There should be an item available now, but another reader might grab it,
+                    // so we keep TryReading in the loop.
                 }
 
                 try
                 {
                     var task = handler(next);
-                    if (task != null && task.Status != TaskStatus.RanToCompletion) await task.ConfigureAwait(false);
+                    if (task != null && task.Status != TaskStatus.RanToCompletion)
+                    {
+                        await task.ConfigureAwait(false);
+                    }
                 }
                 catch { } // matches MessageCompletable
             }
-        }
 
-        internal void UnsubscribeImpl(Exception error = null, CommandFlags flags = CommandFlags.None)
-        {
-            var parent = _parent;
-            _parent = null;
-            if (parent != null)
+            if (Completion.IsFaulted)
             {
-                parent.UnsubscribeAsync(Channel, HandleMessage, flags);
+                _onInternalError(Completion.Exception.InnerException);
             }
-            _queue.Writer.TryComplete(error);
-        }
-
-        internal async Task UnsubscribeAsyncImpl(Exception error = null, CommandFlags flags = CommandFlags.None)
-        {
-            var parent = _parent;
-            _parent = null;
-            if (parent != null)
-            {
-                await parent.UnsubscribeAsync(Channel, HandleMessage, flags).ConfigureAwait(false);
-            }
-            _queue.Writer.TryComplete(error);
         }
 
         internal static bool IsOneOf(Action<RedisChannel, RedisValue> handler)
         {
-            try
-            {
-                return handler?.Target is ChannelMessageQueue
-                    && handler.Method.Name == nameof(HandleMessage);
-            }
-            catch
-            {
-                return false;
-            }
+            // REVIEW: Need more context here to properly replace this.
```

Comment:
I think this can be easily restored by moving it to RedisSubscriber, since that's where the handler lives now.